### ISSUE TITLE
Regex ^.*$ doesn't match lines that end with an incomplete UTF-8 character

### ISSUE THREAD

**Issue by e6c31d** [fluent/fluent-bit#6130]
## Bug Report

**Describe the bug**
If a line ends with an incomplete UTF-8 sequence, the regex `^.*$` doesn't match it.
An example of an incomplete UTF-8 sequence is `\xc3`, which is the first byte of the sequence `\xc3\xb6` (ö). Basically the "leading bytes" in [this table](https://en.wikipedia.org/wiki/UTF-8#Codepage_layout) between `\xc2` and `\xf4` inclusive.

Notes:
1. Incomplete UTF-8 sequences at the beginning or middle of a line do not trigger this bug.
2. Invalid UTF-8 bytes (such as a continuation byte by itself or `\xff`) do not trigger this bug.

**To Reproduce**
- Rubular link if applicable: Couldn't reproduce in Rubular
- Example log message if applicable:
```
printf $'{"log":"abc\xc3"}'
```
- Steps to reproduce the problem:
This one line demonstrates the problem. The second log line does not get matched by `^.*$`:
```
$ printf $'{"log":"abc\xc3def"}{"log":"abc\xc3"}{"log":"\xc3def"}' | docker run --interactive --rm cr.fluentbit.io/fluent/fluent-bit -i stdin -F grep -p 'regex=log ^.*$' -m '*' -o stdout 2>/dev/null
[0] stdin.0: [1664699378.548651700, {"log"=>"abc�def"}]
[1] stdin.0: [1664699378.548660400, {"log"=>"�def"}]
```

**Expected behavior**
The second log line should be matched by `^.*$` even if it ends with `\xc3`:
```
[0] stdin.0: [1664699378.548651700, {"log"=>"abc�def"}]
[1] stdin.0: [1664699378.548652000, {"log"=>"abc�"}]
[2] stdin.0: [1664699378.548660400, {"log"=>"�def"}]
```

**Screenshots**

**Your Environment**

* Version used: latest fluent-bit docker image (Fluent Bit v1.9.9)
* Configuration: See command line in "Steps to reproduce" above.
* Environment name and version: Docker 20.10.17
* Server type and version: latest fluent-bit docker image
* Operating System and version: latest fluent-bit docker image
* Filters and plugins: Grep filter

**Additional context**
I found this issue because containerd seems to split lines longer than 16K bytes regardless of the actual characters, so it sometimes splits a valid UTF-8 character in half, and the first line gets missed by Fluent Bit.

**Comment by nokute78** [comment]
The issue may relate docker.

I tested v2.0.5 without docker and it worked.
```
$ printf $'{"log":"abc\xc3def"}{"log":"abc\xc3"}{"log":"\xc3def"}' | ~/git/WORKTREE/v2.0.5/build/bin/fluent-bit -i stdin -F grep -p 'regex log ^.*$' -m '*' -o stdout
Fluent Bit v2.0.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/12 14:25:07] [ info] [fluent bit] version=2.0.5, commit=8e5f9f9106, pid=34309
[2022/11/12 14:25:07] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/12 14:25:07] [ info] [cmetrics] version=0.5.7
[2022/11/12 14:25:07] [ info] [ctraces ] version=0.2.5
[2022/11/12 14:25:07] [ info] [input:stdin:stdin.0] initializing
[2022/11/12 14:25:07] [ info] [input:stdin:stdin.0] storage_strategy='memory' (memory only)
[2022/11/12 14:25:07] [ info] [sp] stream processor started
[2022/11/12 14:25:07] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2022/11/12 14:25:07] [ warn] [engine] service will shutdown in max 5 seconds
[2022/11/12 14:25:07] [ info] [output:stdout:stdout.0] worker #0 started
[0] stdin.0: [1668230707.872072503, {"log"=>"abc�def"}]
[1] stdin.0: [1668230707.872075919, {"log"=>"abc�"}]
[2] stdin.0: [1668230707.872077041, {"log"=>"�def"}]
[2022/11/12 14:25:08] [ info] [engine] service has stopped (0 pending tasks)
[2022/11/12 14:25:08] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/11/12 14:25:08] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

On the other hand, v2.0.5 with docker didn't work.
```
# printf $'{"log":"abc\xc3def"}{"log":"abc\xc3"}{"log":"\xc3def"}' | docker run --interactive --rm fluent/fluent-bit:2.0.5 -i stdin -F grep -p 'regex=log ^.*$' -m '*' -o stdout
Fluent Bit v2.0.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/12 05:07:49] [ info] [fluent bit] version=2.0.5, commit=8e5f9f9106, pid=1
[2022/11/12 05:07:49] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/12 05:07:49] [ info] [cmetrics] version=0.5.7
[2022/11/12 05:07:49] [ info] [ctraces ] version=0.2.5
[2022/11/12 05:07:49] [ info] [input:stdin:stdin.0] initializing
[2022/11/12 05:07:49] [ info] [input:stdin:stdin.0] storage_strategy='memory' (memory only)
[2022/11/12 05:07:49] [ info] [sp] stream processor started
[2022/11/12 05:07:49] [ warn] [input:stdin:stdin.0] end of file (stdin closed by remote end)
[2022/11/12 05:07:49] [ warn] [engine] service will shutdown in max 5 seconds
[2022/11/12 05:07:49] [ info] [output:stdout:stdout.0] worker #0 started
[0] stdin.0: [1668229669.147545946, {"log"=>"abc�def"}]
[1] stdin.0: [1668229669.147550136, {"log"=>"�def"}]
[2022/11/12 05:07:49] [ info] [engine] service has stopped (0 pending tasks)
[2022/11/12 05:07:49] [ info] [output:stdout:stdout.0] thread worker #0 stopping...
[2022/11/12 05:07:49] [ info] [output:stdout:stdout.0] thread worker #0 stopped
```

**Comment by leonardo-albertovich** [comment]
I would either use a file or tcp input instead just to be able to rule out dockers stdin forwarding and determine if it's something related to fluent-bit running inside docker or not.

**Comment by e6c31d** [comment]
I tried the latest release (v.2.0.5) on Windows 10 using a file as input (see [input.txt](https://github.com/fluent/fluent-bit/files/9996565/input.txt)), and got the same wrong result. The below is run using `Git Bash`.

```
$ printf $'abc\xc3def\nabc\xc3\n\xc3def\n' >./input.txt

$ ./fluent-bit.exe -i tail -p Path=./input.txt -p Read_from_Head=true -F grep -p 'regex=log ^.*$' -m '*' -o stdout
Fluent Bit v2.0.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/13 08:29:42] [ info] [fluent bit] version=2.0.5, commit=8e5f9f9106, pid=24408
[2022/11/13 08:29:42] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/13 08:29:42] [ info] [cmetrics] version=0.5.7
[2022/11/13 08:29:42] [ info] [ctraces ] version=0.2.5
[2022/11/13 08:29:42] [ info] [input:tail:tail.0] initializing
[2022/11/13 08:29:42] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2022/11/13 08:29:42] [ info] [sp] stream processor started
[2022/11/13 08:29:42] [ info] [output:stdout:stdout.0] worker #0 started
[0] tail.0: [1668317382.353868700, {"log"=>"abc▒def"}]
[1] tail.0: [1668317382.353880700, {"log"=>"▒def"}]
```

Also ran it from `cmd.exe`. Same issue:
```
C:\Users\xxx\Desktop\fluent-bit-2.0.5-win64\bin>.\fluent-bit.exe -i tail -p Path=./input.txt -p Read_from_Head=true -F grep -p "regex=log ^.*$" -m "*" -o stdout
Fluent Bit v2.0.5
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2022/11/13 08:33:49] [ info] [fluent bit] version=2.0.5, commit=8e5f9f9106, pid=22748
[2022/11/13 08:33:49] [ info] [storage] ver=1.3.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2022/11/13 08:33:49] [ info] [cmetrics] version=0.5.7
[2022/11/13 08:33:49] [ info] [ctraces ] version=0.2.5
[2022/11/13 08:33:49] [ info] [input:tail:tail.0] initializing
[2022/11/13 08:33:49] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2022/11/13 08:33:49] [ info] [sp] stream processor started
[2022/11/13 08:33:49] [ info] [output:stdout:stdout.0] worker #0 started
[0] tail.0: [1668317629.950408200, {"log"=>"abc├def"}]
[1] tail.0: [1668317629.950427600, {"log"=>"├def"}]
```

**Comment by e6c31d** [comment]
@nokute78 
> $ printf $'{"log":"abc\xc3def"}{"log":"abc\xc3"}{"log":"\xc3def"}' | ~/git/WORKTREE/v2.0.5/build/bin/fluent-bit -i stdin -F grep -p '<b>regex log ^.*$</b>' -m '*' -o stdout

You're missing the `=` between `regex` and `log`. If you add it, you'll get the wrong result.

**Comment by nokute78** [comment]
@e6c31d Oops. Thank you for pointing it.
I can reproduce the issue without docker.

**Comment by nokute78** [comment]
Note: `onig_search` returns `ONIG_MISMATCH`.

If a configuration of regex is changed like below, this issue was fixed.
We may need a fallback to handle string as an ASCII.
```diff
diff --git a/src/flb_regex.c b/src/flb_regex.c
index dff09ce2b..2cc14df18 100644
--- a/src/flb_regex.c
+++ b/src/flb_regex.c
@@ -77,7 +77,7 @@ static int str_to_regex(const char *pattern, OnigRegex *reg)
     ret = onig_new(reg,
                    (const unsigned char *)start, (const unsigned char *)end,
                    ONIG_OPTION_DEFAULT,
-                   ONIG_ENCODING_UTF8, ONIG_SYNTAX_RUBY, &einfo);
+                   ONIG_ENCODING_ASCII, ONIG_SYNTAX_RUBY, &einfo);
 
     if (ret != ONIG_NORMAL) {
         return -1;
```

**Comment by e6c31d** [comment]
This seems to be fixed in 2.0.9.
